llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -1409,6 +1409,19 @@ Instruction *InstCombinerImpl::foldICmpTruncConstant(ICmpInst &Cmp,
                                                      const APInt &C) {
   ICmpInst::Predicate Pred = Cmp.getPredicate();
   Value *X = Trunc->getOperand(0);
+  Type *SrcTy = X->getType();
+  unsigned DstBits = Trunc->getType()->getScalarSizeInBits(),
+           SrcBits = SrcTy->getScalarSizeInBits();
+
+  // Match (icmp pred (trunc nuw/nsw X), C)
+  // Which we can convert to (icmp pred X, (sext/zext C))
+  if (shouldChangeType(DstBits, SrcBits)) {
```

Comment:
Shouldn't this be passing the types, not the bit width? That way we will not transform vector cases, for which we don't have sensible profitability heuristics here.

Reply:
Sure, in this case the `shouldChnageType` that accepts types just forwards the `primitiveBitWidth` to the bitwidth helper, but if you think this version is a more future proof happy to change.

Reply:
This isn't really true for vectors. If the previous trunc + icmp sequence worked on legal types and you convert it into one that works on illegal types, that may be catastrophically worse (down to scalarization). There's a reason the special handling for vectors is there.

Reply:
Here's a trivial example for aarch64: https://llvm.godbolt.org/z/K3hh3bx65 Even in the one-use case, getting rid of the trunc is worse for codegen. I haven't checked, but I expect your patch would currently do this transform because i32 is legal but i8 is not (for scalars).

Reply:
Updated.